dist/restore/index.js

```diff
@@ -0,0 +1,168 @@
+    const gcsKeyfile = JSON.parse(buffer.toString('utf8'));
+
+    // Create a GCS client
+    const storage = await new Storage({ credentials: gcsKeyfile });
+    const bucketName = gcsKeyfile.project_id;
+    const bucket = await storage.bucket(bucketName);
+
+    // Get all caches available in Bucket for the given cache key
+    const [availableCaches] = await bucket.getFiles({
+      prefix: cacheKey
+    });
+
+    if (typeof availableCaches !== undefined && availableCaches.length > 0) {
+      console.log(`Caches for key ${cacheKey} found in bucket ${bucketName}`);
+    } else {
+      // Exit if no caches are found
+      console.log(`No caches found for key ${cacheKey} in the bucket ${bucketName}`);
+      return;
+    }
+
+    // List all given path with their corresponding cache references (if exists in the bucket)
+    const cacheObjects = pathsString.split('\n').map((cachePath) => {
+      const cacheFileName = fileNameForCache(cacheKey, cachePath);
+      const cacheReference = availableCaches.find((cache) => {
+        return cache.name == cacheFileName;
+      });
+
+      return [cachePath, cacheReference];
+    });
+
+    const foundCachesMask = cacheObjects.map((cachePath) => {
+      return cachePath[1] !== undefined;
+    });
+
+    core.setOutput('cache-hit', foundCachesMask.every((hasCache) => { return hasCache === true } ));
+
+    // TODO: Check if it is possible to send the paths of the missing caches to the Post Action via environment variable
+    // process.env.NOT_FOUND_CACHES = '[]';
+
+    const filteredCacheObjects = cacheObjects.filter((_, index) => { return foundCachesMask[index] } );
+
+    if (typeof filteredCacheObjects !== undefined && filteredCacheObjects.length > 0) {
+      if (noDownload === 'false') {
+        // Start a download for each cache asynchronously
+        await Promise.all(filteredCacheObjects.map(async ([path, reference]) => {
+            return await downloadCache(path, reference);
+          }));
+      }
+    }
+  }
+  catch (error) {
+    core.setFailed(error.message);
+  }
+}
+
+function fileNameForCache(key, cachePath) {
+  return [key, '_', md5(cachePath), compressedFileExtension].join('');
+}
+
+async function downloadCache(cachePath, gcsFileObject) {
+  // Set names for temporary download folder and target folder
+  const downloadFolder = [
+      temporaryDirectoryPrefix,
+      gcsFileObject.name.replace(compressedFileExtension, '')
+    ].join('_');
+  const cacheFolderPath = path.dirname(cachePath);
+
+  // TODO delete it
+  console.log(downloadFolder);
+  console.log(cacheFolderPath);
+
+  const download = await gcsFileObject.exists()
+    .then(async (response) => {
+      const exists = response[0];
+      if (!exists) { throw new Error(`${cachePath} could not be restored`) }
+
+      function createFolder(folderPath) {
+        if (!fs.existsSync(cacheFolderPath)) {
+          fs.mkdirSync(cacheFolderPath, { recursive: true });
+        }
+      }
```

Comment:
Use guarding pattern instead
```suggestion
      function createFolder(folderPath) {
        if (fs.existsSync(cacheFolderPath)) { return }
        fs.mkdirSync(cacheFolderPath, { recursive: true });
      }
```